pkg/controller/unifiedpushserver/backups.go

```diff
@@ -0,0 +1,130 @@
+package unifiedpushserver
+
+import (
+	"fmt"
+
+	pushv1alpha1 "github.com/aerogear/unifiedpush-operator/pkg/apis/push/v1alpha1"
+	batchv1 "k8s.io/api/batch/v1"
+	batchv1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func backups(ups *pushv1alpha1.UnifiedPushServer) ([]batchv1beta1.CronJob, error) {
+	cronjobs := []batchv1beta1.CronJob{}
+	for _, upsBackup := range ups.Spec.Backups {
+		cronJobLabels := labels(ups, "backup")
+		for k, v := range upsBackup.Labels {
+			cronJobLabels[k] = v
+		}
+
+		jobLabels := map[string]string{"cronjob-name": upsBackup.Name}
+		for k, v := range cronJobLabels {
+			jobLabels[k] = v
+		}
+
+		cronjobs = append(cronjobs, batchv1beta1.CronJob{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      upsBackup.Name,
+				Namespace: ups.Namespace,
+				Labels:    cronJobLabels,
+			},
+			Spec: batchv1beta1.CronJobSpec{
+				Schedule: upsBackup.Schedule,
+				JobTemplate: batchv1beta1.JobTemplateSpec{
+					Spec: batchv1.JobSpec{
+						Template: corev1.PodTemplateSpec{
+							ObjectMeta: metav1.ObjectMeta{
+								Labels: jobLabels,
+							},
+							Spec: corev1.PodSpec{
+								// This SA needs to be created beforehand
+								// https://github.com/integr8ly/backup-container-image/tree/master/templates/openshift/rbac
+								ServiceAccountName: "backupjob",
+								Containers: []corev1.Container{
+									{
+										Name:            upsBackup.Name + "-keycloak-backup",
+										Image:           cfg.BackupImage,
+										ImagePullPolicy: "Always",
+										Command:         buildBackupContainerCommand(upsBackup, ups.Namespace),
+										Env:             buildBackupCronJobEnvVars(upsBackup, ups.Name, ups.Namespace),
+									},
+								},
+								RestartPolicy: corev1.RestartPolicyOnFailure,
+							},
+						},
+					},
+				},
+			},
+		})
+	}
+	return cronjobs, nil
+}
+
+func buildBackupContainerCommand(upsBackup pushv1alpha1.UnifiedPushServerBackup, upsNamespace string) []string {
+	command := []string{"/opt/intly/tools/entrypoint.sh", "-c", "postgres", "-n", upsNamespace}
+
+	// If there is no encryption secret, we need to inhibit the
+	// encryption behaviour
+	if upsBackup.EncryptionKeySecretName == "" {
+		command = append(command, "-e", "")
+	}
+
+	return command
+}
+
+func buildBackupCronJobEnvVars(upsBackup pushv1alpha1.UnifiedPushServerBackup, upsName string, upsNamespace string) []corev1.EnvVar {
+
+	envVars := []corev1.EnvVar{
+		{
+			Name:  "PRODUCT_NAME",
+			Value: "unifiedpush",
+		},
+		{
+			Name:  "COMPONENT_SECRET_NAME",
+			Value: fmt.Sprintf("%s-%s", upsName, "postgresql"),
```

Comment:
can you add ""%s-%s-secret" as it is used in other products?

Reply:
We could, but it seems redundant to me to add the kind to the name :thinking: Is there a benefit to doing that?

Do we add `-deploymentconfig` to the end of DeploymentConfig names? Would you add `-string` to the end of a String value in a typed programming language?

Reply:
I agree with you but that's how other component secrets are being created today (for now at least)

Reply:
That's good to know. Do you think it will cause us any problems if we don't change?

Reply:
not at all, "just a convention", lets keep as it is

Reply:
Are secret names limited to 63 characters?

Reply:
all resources names (and labels) are limited to 63 chars: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set

Reply:
@grdryn if someone changes the ENV VARs values manually by OCP console it will stop to work. To avoid this scenario in the MSS Oper we are checking the ENV VAR values from the database pod as the service name dynamically and creating all programmatically. 

c/c @odra

Reply:
This is a good point. I think if we can work out (based on the suffixes that we append) what max string lengths for names that we can have that we should add to our custom kinds. I think adding kubebuilder validation annotations should cause ones that are over that to be rejected, but I'm not sure if that's enabled in OpenShift yet or not.

Reply:
@grdryn the database env vars (user/pwd/database name) are env variables of the pod/deployment usually. I think it is in this way here which means that a user can change it for someone reason. 

In MSS oper we fetch the pod database then get the env vars used. You can check here we building the DB secret data . https://github.com/aerogear/mobile-security-service-operator/pull/125/files#diff-0a376a010c2ec86f696671db622a8955R12

Reply:
@camilamacedo86 yes, for the container they're passed as env vars, but those env var values are taken from the secret, so I don't think we suffer from the issue that you're describing. Also, since it's just one controller for all of this, it should be kept in sync, even if someone does change something.

Here's where we create the secret:
https://github.com/aerogear/unifiedpush-operator/blob/076d019ad3c21a3ed82ad505e158e51f71aa275b/pkg/controller/unifiedpushserver/postgresql.go#L41..L50

Here's where we link the secret values to the container env vars: https://github.com/aerogear/unifiedpush-operator/blob/076d019ad3c21a3ed82ad505e158e51f71aa275b/pkg/controller/unifiedpushserver/postgresql.go#L88..L122

Reply:
Hi @grdryn,

After the installation via the UPS oper, all will be fine. However, if the admin changed it in OCP for any reason the backup will stop to work. IMHO: We need to think that how much each avoids issues is better and that it will be used now for the RHMI managed by us and self-managed as well. Also, by the community which should be able to use just the UPS oper alone to get the service as well.